stl/inc/ranges

```diff
@@ -6842,6 +6843,472 @@ namespace ranges {
+        requires(sizeof...(_LHSTupleTypes) == sizeof...(_RHSTupleTypes))
+    _NODISCARD constexpr bool _Zip_iterator_sentinel_equal(
+        const tuple<_LHSTupleTypes...>& _Lhs_tuple, const tuple<_RHSTupleTypes...>& _Rhs_tuple)
+        // clang-format off
+        noexcept((noexcept(_STD declval<const _LHSTupleTypes&>() == _STD declval<const _RHSTupleTypes&>()) && ...)) {
+        // clang-format on
+        const auto _Evaluate_equality_closure =
+            [&_Lhs_tuple, &_Rhs_tuple ]<size_t... _Indices>(index_sequence<_Indices...>) {
+            return ((_STD get<_Indices>(_Lhs_tuple) == _STD get<_Indices>(_Rhs_tuple)) || ...);
+        };
+
+        return _Evaluate_equality_closure(index_sequence_for<_LHSTupleTypes...>{});
+    }
+
+    // clang-format off
+    _EXPORT_STD template <input_range... _ViewTypes>
+        requires (view<_ViewTypes> && ...) && (sizeof...(_ViewTypes) > 0)
+    class zip_view : public view_interface<zip_view<_ViewTypes...>> {
+        // clang-format on
+    private:
+        /* [[no_unique_address]] */ tuple<_ViewTypes...> _Views;
+
+        template <bool _IsConst>
+        struct _Category_base {};
+
+        template <bool _IsConst>
+            requires _All_forward<_IsConst, _ViewTypes...>
+        struct _Category_base<_IsConst> {
+            using iterator_category = input_iterator_tag;
+        };
+
+        template <bool _IsConst>
+        class _Iterator : public _Category_base<_IsConst> {
+        private:
+            /* [[no_unique_address]] */ tuple<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>...> _Current;
+
+            friend zip_view;
+
+            constexpr explicit _Iterator(tuple<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>...> _Current_) noexcept(
+                is_nothrow_move_constructible_v<
+                    tuple<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>...>>) // strengthened
+                : _Current(_STD move(_Current_)) {}
+
+        public:
+            using iterator_concept = conditional_t<_All_random_access<_IsConst, _ViewTypes...>,
+                random_access_iterator_tag,
+                conditional_t<_All_bidirectional<_IsConst, _ViewTypes...>, bidirectional_iterator_tag,
+                    conditional_t<_All_forward<_IsConst, _ViewTypes...>, forward_iterator_tag, input_iterator_tag>>>;
+
+            using value_type      = tuple<range_value_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+            using difference_type = common_type_t<range_difference_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+
+            _Iterator() = default;
+
+            constexpr _Iterator(_Iterator<!_IsConst> _Rhs) noexcept(
+                (is_nothrow_convertible_v<iterator_t<_ViewTypes>,
+                     iterator_t<_Maybe_const<!_IsConst, _ViewTypes>>> && ...)) // strengthened
+                // clang-format off
+                requires(_IsConst &&
+            (convertible_to<iterator_t<_ViewTypes>, iterator_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...))

```

Comment:
@CaseyCarter noticed that the Standardese says `(convertible_to<iterator_t<Views>, iterator_t<const Views>> && ...)`. The constraint has already guaranteed that `_IsConst` is true, so using `_Maybe_const<_IsConst,` is correct, but unnecessarily verbose.

Reply:
You know, that's been a pretty interesting issue for me. The original paper P2321R2 also used `iterator_t<maybe-const<Const, Views>>` instead of just `iterator_t<const Views>`, but that's no longer how it is written in the C++ Working Draft, as you mentioned. There, they do indeed directly use `iterator_t<const Views>`.

In theory, this shouldn't affect it, as you mentioned, but something odd is going on. A similar situation occurs with `sentinel_t<maybe-const<Const, Views>>` instead of `sentinel_t<const Views>`, but when using `sentinel_t<const Views>` instead of `sentinel_t<maybe-const<Const, Views>>`, the following libc++ test fails to compile in Clang:

```C++
// From std/ranges/range.adaptors/range.zip/sentinel/ctor.default.pass.cpp:

struct PODSentinel {
	bool b; // deliberately uninitialised

	friend constexpr bool operator==(int*, const PODSentinel& s) { return s.b; }
};

struct Range : std::ranges::view_base {
	int* begin() const;
	PODSentinel end();  // const Range does not satisfy std::ranges::range.
};

constexpr bool test() {
	using Iterator_T = std::ranges::iterator_t<Range>;

	{
		using R = std::ranges::zip_view<Range>;

		// With sentinel_t<maybe-const<Const, Views>>: Program Compiles
		// With sentinel_t<const Views>: Constraints for std::ranges::sentinel_t<R> not satisfied because
		// (convertible_to<sentinel_t<Views>, sentinel_t<const Views>> && ...) is not satisfied
		using Sentinel = std::ranges::sentinel_t<R>;
		
		// ...
	}
	return true;
}

// ...
```

It's weird, though, because I can't seem to find any LWG issue or P2321 revision which changed the use of `iterator_t<maybe-const<Const, Views>>` to `iterator_t<const Views>`. Since I'm working on checking the libc++ `zip_view` tests for this implementation, I'm just going to mark this test and `std\ranges\range.adaptors\range.zip\begin.pass.cpp` as `FAIL`, since they both suffer from this same underlying problem. Who knows? Maybe the tests were borked from the start...

Reply:
> The original paper P2321R2 also used `iterator_t<maybe-const<Const, Views>>` instead of just `iterator_t<const Views>`, but that's no longer how it is written in the C++ Working Draft, as you mentioned.

For the record, this is https://github.com/cplusplus/draft/pull/5778

Clang is known to be too eager in instantiating types in constraints. I believe that this is fixed for Clang 16.

Reply:
Ah, that makes a lot more sense now! I was starting to go crazy during testing, but it's nice to know that the problem isn't on my side this time. I'm also still fairly new to the STL implementation scene, so I wasn't aware that minor changes to the C++ Working Draft could be made as PRs there. Thank you for the update!